File: phases/13-concurrent-and-parallel/03-memory-models-sequential-consistency-vs-relaxed/code/main.cpp
// Memory Models — Sequential Consistency vs Relaxed
// Phase 13 — Concurrent & Parallel Computing
// Litmus test suite demonstrating SC, TSO, and relaxed ordering
//
// Compile: g++ -O2 -pthread -std=c++20 -o memmodel main.cpp && ./memmodel

#include <atomic>
#include <thread>
#include <iostream>
#include <cassert>
#include <chrono>
#include <cstring>

// =============================================================================
// Part 1: Sequentially Consistent Atomics — Dekker Pattern
// Under seq_cst, (r1=0, r2=0) is impossible because a global total order
// guarantees at least one thread sees the other's write.
// =============================================================================

void part1_dekker_sc() {
    std::cout << "=== Part 1: SC Dekker Pattern ===\n";
    constexpr int ITERATIONS = 50000;

    std::atomic<int> x{0};
    std::atomic<int> y{0};
    int r1{0}, r2{0};

    int outcomes[4] = {0};

    for (int i = 0; i < ITERATIONS; ++i) {
        x.store(0, std::memory_order_seq_cst);
        y.store(0, std::memory_order_seq_cst);
        r1 = r2 = 0;

        std::thread t1([&]() {
            x.store(1, std::memory_order_seq_cst);
            r1 = y.load(std::memory_order_seq_cst);
        });

        std::thread t2([&]() {
            y.store(1, std::memory_order_seq_cst);
            r2 = x.load(std::memory_order_seq_cst);
        });

        t1.join();
        t2.join();

        outcomes[r1 * 2 + r2]++;
    }

    std::cout << "SC Dekker outcomes (" << ITERATIONS << " runs):\n";
    std::cout << "  (0,0): " << outcomes[0] << "  (should be 0)\n";
    std::cout << "  (0,1): " << outcomes[1] << "\n";
    std::cout << "  (1,0): " << outcomes[2] << "\n";
    std::cout << "  (1,1): " << outcomes[3] << "\n";

    if (outcomes[0] == 0) {
        std::cout << "  ✓ SC guarantee holds: (0,0) never observed\n";
    } else {
        std::cout << "  ⚠ Unexpected (0,0) observed — hardware weaker than SC?\n";
    }
}

// =============================================================================
// Part 2: Relaxed Atomics — Surprising Results
// With memory_order_relaxed, the compiler and CPU can reorder stores.
// The reader can observe (a=1, b=0) — b written after a but visible first.
// =============================================================================

void part2_relaxed_reordering() {
    std::cout << "\n=== Part 2: Relaxed Atomics — Surprising Reordering ===\n";
    constexpr int ITERATIONS = 100000;

    std::atomic<int> a{0};
    std::atomic<int> b{0};
    int observed[4] = {0};

    for (int iter = 0; iter < ITERATIONS; ++iter) {
        a.store(0, std::memory_order_relaxed);
        b.store(0, std::memory_order_relaxed);

        std::thread writer([&]() {
            a.store(1, std::memory_order_relaxed);
            b.store(1, std::memory_order_relaxed);
        });

        std::thread reader([&]() {
            int i = b.load(std::memory_order_relaxed);
            int j = a.load(std::memory_order_relaxed);
            observed[i * 2 + j]++;
        });

        writer.join();
        reader.join();
    }

    std::cout << "Relaxed outcomes (" << ITERATIONS << " runs):\n";
    std::cout << "  (a=0,b=0): " << observed[0] << "\n";
    std::cout << "  (a=0,b=1): " << observed[1] << "\n";
    std::cout << "  (a=1,b=0): " << observed[2]
              << "  (reordered! b=1 visible before a=1)\n";
    std::cout << "  (a=1,b=1): " << observed[3] << "\n";

    double pct = 100.0 * observed[2] / ITERATIONS;
    std::cout << "  → " << pct << "% of runs showed reordering\n";
}

// =============================================================================
// Part 3: Message Passing — Acquire/Release vs Relaxed
// Producer stores data, then sets ready flag.
// Consumer spins on ready, then reads data.
//
// With acquire/release: data is always visible when ready is set.
// With relaxed: the assertion can fail — the consumer sees ready=true
// but data is still 0.
// =============================================================================

void part3_message_passing_relaxed() {
    std::cout << "\n=== Part 3a: Message Passing with Relaxed (should fail) ===\n";
    constexpr int ITERATIONS = 50000;

    std::atomic<bool> ready{false};
    int data{0};
    int failures = 0;

    for (int i = 0; i < ITERATIONS && failures < 5; ++i) {
        data = 0;
        ready.store(false, std::memory_order_relaxed);

        std::thread producer([&]() {
            data = 42;
            ready.store(true, std::memory_order_relaxed);
        });

        std::thread consumer([&]() {
            while (!ready.load(std::memory_order_relaxed)) {}
            if (data != 42) {
                failures++;
            }
        });

        producer.join();
        consumer.join();
    }

    if (failures > 0) {
        std::cout << "  ⚠ Failed " << failures << " times: relaxed allows data race\n";
    } else {
        std::cout << "  No failures observed on this hardware (common on x86)\n";
    }
}

void part3_message_passing_acq_rel() {
    std::cout << "\n=== Part 3b: Message Passing with Acquire/Release (always correct) ===\n";
    constexpr int ITERATIONS = 50000;

    std::atomic<bool> ready{false};
    int data{0};
    int failures = 0;

    for (int i = 0; i < ITERATIONS; ++i) {
        data = 0;
        ready.store(false, std::memory_order_release);

        std::thread producer([&]() {
            data = 42;
            ready.store(true, std::memory_order_release);
        });

        std::thread consumer([&]() {
            while (!ready.load(std::memory_order_acquire)) {}
            if (data != 42) {
                failures++;
            }
        });

        producer.join();
        consumer.join();
    }

    if (failures == 0) {
        std::cout << "  ✓ Acquire/release guarantees: data always visible (0 failures)\n";
    } else {
        std::cout << "  ⚠ Unexpected failures: " << failures << "\n";
    }
}

// =============================================================================
// Part 4: Memory Fences
// Using std::atomic_thread_fence instead of ordering on atomic operations.
// A release fence + relaxed store has the same effect as a release store.
// =============================================================================

void part4_fences() {
    std::cout << "\n=== Part 4: Memory Fences ===\n";
    constexpr int ITERATIONS = 50000;

    std::atomic<bool> ready{false};
    int data{0};
    int failures = 0;

    for (int i = 0; i < ITERATIONS; ++i) {
        data = 0;
        ready.store(false, std::memory_order_relaxed);

        std::thread producer([&]() {
            data = 42;
            std::atomic_thread_fence(std::memory_order_release);
            ready.store(true, std::memory_order_relaxed);
        });

        std::thread consumer([&]() {
            while (!ready.load(std::memory_order_relaxed)) {}
            std::atomic_thread_fence(std::memory_order_acquire);
            if (data != 42) {
                failures++;
            }
        });

        producer.join();
        consumer.join();
    }

    if (failures == 0) {
        std::cout << "  ✓ Release fence + acquire fence: always correct (0 failures)\n";
    } else {
        std::cout << "  ⚠ Failures with fences: " << failures << "\n";
    }
}

// =============================================================================
// Part 5: IRIW Litmus Test (Independent Reads of Independent Writes)
// Two threads each write to their own variable.
// Two reader threads observe both variables.
// Under SC, all readers should agree on the order of the two writes.
// On relaxed hardware (ARM/Power), readers can disagree.
// =============================================================================

void part5_iriw() {
    std::cout << "\n=== Part 5: IRIW (Independent Reads of Independent Writes) ===\n";
    constexpr int ITERATIONS = 50000;

    std::atomic<int> x{0};
    std::atomic<int> y{0};
    int disagreement_count = 0;

    for (int i = 0; i < ITERATIONS; ++i) {
        x.store(0, std::memory_order_seq_cst);
        y.store(0, std::memory_order_seq_cst);
        int r1 = 0, r2 = 0, r3 = 0, r4 = 0;

        std::thread t1([&]() {
            x.store(1, std::memory_order_seq_cst);
        });

        std::thread t2([&]() {
            y.store(1, std::memory_order_seq_cst);
        });

        std::thread t3([&]() {
            r1 = x.load(std::memory_order_seq_cst);
            r2 = y.load(std::memory_order_seq_cst);
        });

        std::thread t4([&]() {
            r3 = y.load(std::memory_order_seq_cst);
            r4 = x.load(std::memory_order_seq_cst);
        });

        t1.join();
        t2.join();
        t3.join();
        t4.join();

        // Reader 1 saw (x=1, y=0) and reader 2 saw (y=1, x=0) → disagreement
        if (r1 == 1 && r2 == 0 && r3 == 1 && r4 == 0) {
            disagreement_count++;
        }
    }

    if (disagreement_count > 0) {
        std::cout << "  Disagreements: " << disagreement_count
                  << " (hardware weaker than SC)\n";
    } else {
        std::cout << "  ✓ IRIW: no disagreements, hardware is SC or TSO\n";
    }
}

// =============================================================================
// Main — Run all litmus tests
// =============================================================================

int main() {
    std::cout << "Memory Model Litmus Test Suite\n";
    std::cout << "==============================\n";
    std::cout << "Hardware: ";

#ifdef __x86_64__
    std::cout << "x86-64 (TSO)\n\n";
#elif defined(__aarch64__)
    std::cout << "ARM64 (Relaxed)\n\n";
#else
    std::cout << "Unknown\n\n";
#endif

    part1_dekker_sc();
    part2_relaxed_reordering();
    part3_message_passing_relaxed();
    part3_message_passing_acq_rel();
    part4_fences();
    part5_iriw();

    std::cout << "\nDone. See docs/en.md for explanation of each result.\n";
    return 0;
}

File: phases/13-concurrent-and-parallel/03-memory-models-sequential-consistency-vs-relaxed/code/main.rs
//! Memory Models — Sequential Consistency vs Relaxed
//! Phase 13 — Concurrent & Parallel Computing
//!
//! Litmus test suite in Rust demonstrating sequential consistency,
//! relaxed ordering, acquire/release, and memory fences.
//!
//! Run: rustc -O main.rs && ./main

use std::sync::atomic::{AtomicBool, AtomicIsize, Ordering};
use std::sync::Arc;
use std::thread;

// =============================================================================
// Part 1: Sequential Consistency — Dekker Pattern
// =============================================================================

fn part1_dekker_sc() {
    println!("=== Part 1: SC Dekker Pattern ===");
    let iterations = 50_000;

    let x = Arc::new(AtomicIsize::new(0));
    let y = Arc::new(AtomicIsize::new(0));

    let mut outcomes = [0usize; 4];

    for _ in 0..iterations {
        x.store(0, Ordering::SeqCst);
        y.store(0, Ordering::SeqCst);

        let x1 = Arc::clone(&x);
        let y1 = Arc::clone(&y);
        let t1 = thread::spawn(move || {
            x1.store(1, Ordering::SeqCst);
            y1.load(Ordering::SeqCst)
        });

        let x2 = Arc::clone(&x);
        let y2 = Arc::clone(&y);
        let t2 = thread::spawn(move || {
            y2.store(1, Ordering::SeqCst);
            x2.load(Ordering::SeqCst)
        });

        let r1 = t1.join().unwrap();
        let r2 = t2.join().unwrap();
        outcomes[(r1 * 2 + r2) as usize] += 1;
    }

    println!("SC Dekker outcomes ({} runs):", iterations);
    println!("  (0,0): {}  (should be 0)", outcomes[0]);
    println!("  (0,1): {}", outcomes[1]);
    println!("  (1,0): {}", outcomes[2]);
    println!("  (1,1): {}", outcomes[3]);

    if outcomes[0] == 0 {
        println!("  ✓ SC guarantee holds: (0,0) never observed");
    } else {
        println!("  ⚠ Unexpected (0,0) observed");
    }
}

// =============================================================================
// Part 2: Relaxed Atomics — Surprising Reordering
// =============================================================================

fn part2_relaxed_reordering() {
    println!("\n=== Part 2: Relaxed Atomics — Surprising Reordering ===");
    let iterations = 100_000;

    let a = Arc::new(AtomicIsize::new(0));
    let b = Arc::new(AtomicIsize::new(0));
    let mut observed = [0usize; 4];

    for _ in 0..iterations {
        a.store(0, Ordering::Relaxed);
        b.store(0, Ordering::Relaxed);

        let a_w = Arc::clone(&a);
        let b_w = Arc::clone(&b);
        let writer = thread::spawn(move || {
            a_w.store(1, Ordering::Relaxed);
            b_w.store(1, Ordering::Relaxed);
        });

        let a_r = Arc::clone(&a);
        let b_r = Arc::clone(&b);
        let reader = thread::spawn(move || {
            let i = b_r.load(Ordering::Relaxed);
            let j = a_r.load(Ordering::Relaxed);
            (i, j)
        });

        writer.join().unwrap();
        let (i, j) = reader.join().unwrap();
        observed[(i * 2 + j) as usize] += 1;
    }

    println!("Relaxed outcomes ({} runs):", iterations);
    println!("  (a=0,b=0): {}", observed[0]);
    println!("  (a=0,b=1): {}", observed[1]);
    println!("  (a=1,b=0): {}  (reordered!)", observed[2]);
    println!("  (a=1,b=1): {}", observed[3]);

    let pct = 100.0 * observed[2] as f64 / iterations as f64;
    println!("  → {:.1}% of runs showed reordering", pct);
}

// =============================================================================
// Part 3: Message Passing — Acquire/Release
// =============================================================================

fn part3_message_passing() {
    println!("\n=== Part 3: Message Passing with Acquire/Release ===");
    let iterations = 50_000;

    let ready = Arc::new(AtomicBool::new(false));
    let mut failures = 0usize;

    for i in 0..iterations {
        if failures >= 5 {
            println!("  (stopping early after {} failures)", failures);
            break;
        }

        let ready = Arc::new(AtomicBool::new(false));
        let data = Arc::new(AtomicIsize::new(0));

        let r_clone = Arc::clone(&ready);
        let d_clone = Arc::clone(&data);
        let producer = thread::spawn(move || {
            d_clone.store(42, Ordering::Relaxed);
            r_clone.store(true, Ordering::Release);
        });

        let r2 = Arc::clone(&ready);
        let d2 = Arc::clone(&data);
        let consumer = thread::spawn(move || {
            while !r2.load(Ordering::Acquire) {}
            let val = d2.load(Ordering::Relaxed);
            (val, val != 42)
        });

        producer.join().unwrap();
        let (val, failed) = consumer.join().unwrap();
        if failed {
            failures += 1;
            println!("  ⚠ Iteration {}: data = {}", i, val);
        }
    }

    if failures == 0 {
        println!("  ✓ Acquire/release: all iterations passed (0 failures)");
    }
}

// =============================================================================
// Part 4: Relaxed Message Passing (demonstrating the danger)
// =============================================================================

fn part4_message_passing_relaxed() {
    println!("\n=== Part 4: Message Passing with Relaxed (unsafe) ===");
    let iterations = 50_000;

    let mut failures = 0usize;

    for _ in 0..iterations {
        if failures >= 5 {
            break;
        }

        let ready = Arc::new(AtomicBool::new(false));
        let data = Arc::new(AtomicIsize::new(0));

        let r_clone = Arc::clone(&ready);
        let d_clone = Arc::clone(&data);
        let producer = thread::spawn(move || {
            d_clone.store(42, Ordering::Relaxed);
            r_clone.store(true, Ordering::Relaxed);
        });

        let r2 = Arc::clone(&ready);
        let d2 = Arc::clone(&data);
        let consumer = thread::spawn(move || {
            while !r2.load(Ordering::Relaxed) {}
            let val = d2.load(Ordering::Relaxed);
            (val, val != 42)
        });

        producer.join().unwrap();
        let (val, failed) = consumer.join().unwrap();
        if failed {
            failures += 1;
        }
    }

    if failures > 0 {
        println!("  ⚠ Failed {} times with relaxed ordering", failures);
    } else {
        println!("  No failures on this hardware (common on x86)");
    }
}

// =============================================================================
// Part 5: Memory Fences
// =============================================================================

fn part5_fences() {
    println!("\n=== Part 5: Memory Fences ===");
    let iterations = 50_000;

    let ready = Arc::new(AtomicBool::new(false));
    let mut failures = 0usize;

    for _ in 0..iterations {
        let ready = Arc::new(AtomicBool::new(false));
        let data = Arc::new(AtomicIsize::new(0));

        let r_clone = Arc::clone(&ready);
        let d_clone = Arc::clone(&data);
        let producer = thread::spawn(move || {
            d_clone.store(42, Ordering::Relaxed);
            std::sync::atomic::fence(Ordering::Release);
            r_clone.store(true, Ordering::Relaxed);
        });

        let r2 = Arc::clone(&ready);
        let d2 = Arc::clone(&data);
        let consumer = thread::spawn(move || {
            while !r2.load(Ordering::Relaxed) {}
            std::sync::atomic::fence(Ordering::Acquire);
            let val = d2.load(Ordering::Relaxed);
            (val, val != 42)
        });

        producer.join().unwrap();
        let (_, failed) = consumer.join().unwrap();
        if failed {
            failures += 1;
        }
    }

    if failures == 0 {
        println!("  ✓ Fence-based acquire/release: 0 failures");
    } else {
        println!("  ⚠ Fence failures: {}", failures);
    }
}

// =============================================================================
// Main
// =============================================================================

fn main() {
    println!("Memory Model Litmus Test Suite (Rust)");
    println!("=====================================");

    #[cfg(target_arch = "x86_64")]
    println!("Hardware: x86-64 (TSO)\n");
    #[cfg(target_arch = "aarch64")]
    println!("Hardware: ARM64 (Relaxed)\n");

    part1_dekker_sc();
    part2_relaxed_reordering();
    part3_message_passing();
    part4_message_passing_relaxed();
    part5_fences();

    println!("\nDone.");
}
